patches/coreboot-4.8.1/0009-Add-heads-TPM-measurements-to-Skylake-Kabylake.patch

```diff
@@ -0,0 +1,181 @@
+ #include <program_loading.h>
++#include <security/tpm/tss.h>
+ #include <soc/intel/common/util.h>
+ 
+ /*
+@@ -43,3 +44,11 @@ void main(void)
+ 	/* Load and run ramstage. */
+ 	run_ramstage();
+ }
++
++void platform_segment_loaded(uintptr_t start, size_t size, int flags)
++{
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT) && !(flags & SEG_NO_MEASURE)) {
++	  tlcl_measure(2, (const void*) start, size);
++	}
++}
++
+diff --git ./src/drivers/intel/fsp2_0/memory_init.c ./src/drivers/intel/fsp2_0/memory_init.c
+index 30987ce..124d3fa 100644
+--- ./src/drivers/intel/fsp2_0/memory_init.c
++++ ./src/drivers/intel/fsp2_0/memory_init.c
+@@ -150,10 +150,11 @@ static void do_fsp_post_memory_init(bool s3wake, uint32_t fsp_version)
+ 
+ 	/*
+ 	 * Initialize the TPM, unless the TPM was already initialized
+-	 * in verstage and used to verify romstage.
++	 * in verstage and used to verify romstage, or for measured boot.
+ 	 */
+ 	if (IS_ENABLED(CONFIG_LPC_TPM) &&
+-	    !IS_ENABLED(CONFIG_VBOOT_STARTS_IN_BOOTBLOCK))
++	    !IS_ENABLED(CONFIG_VBOOT_STARTS_IN_BOOTBLOCK) &&
++			!IS_ENABLED(CONFIG_MEASURED_BOOT))
+ 		init_tpm(s3wake);
+ }
+ 
+@@ -483,8 +484,33 @@ void fsp_memory_init(bool s3wake)
+ 	if (status != CB_SUCCESS)
+ 		die("Loading FSPM failed!\n");
+ 
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT) && IS_ENABLED(CONFIG_LPC_TPM)) {
++		// we don't know if we are coming out of a resume
++		// at this point, but want to setup the tpm ASAP
++		init_tpm(0);
++		tlcl_lib_init();
++		const void * const bootblock = (const void*) 0xFFFFF800;
++		const unsigned bootblock_size = 0x800;
++		tlcl_measure(0, bootblock, bootblock_size);
++
++		tlcl_measure(1, _romstage, _eromstage - _romstage);
++	}
++
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT)) {
++	  tlcl_measure(1, (const void*) hdr.image_base, hdr.image_size);
++	}
++
+ 	/* Signal that FSP component has been loaded. */
+-	prog_segment_loaded(hdr.image_base, hdr.image_size, SEG_FINAL);
++	// Don't measure since it is relocated at this point
++	prog_segment_loaded(hdr.image_base, hdr.image_size, SEG_FINAL | SEG_NO_MEASURE);
+ 
+ 	do_fsp_memory_init(&hdr, s3wake, &memmap);
+ }
++
++void platform_segment_loaded(uintptr_t start, size_t size, int flags)
++{
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT) && !(flags & SEG_NO_MEASURE)) {
++	  tlcl_measure(1, (const void*) start, size);
++	}
++}
++
+diff --git ./src/drivers/intel/fsp2_0/silicon_init.c ./src/drivers/intel/fsp2_0/silicon_init.c
+index bda88d1..49220af 100644
+--- ./src/drivers/intel/fsp2_0/silicon_init.c
++++ ./src/drivers/intel/fsp2_0/silicon_init.c
+@@ -18,6 +18,7 @@
+ #include <fsp/api.h>
+ #include <fsp/util.h>
+ #include <program_loading.h>
++#include <security/tpm/tss.h>
+ #include <stage_cache.h>
+ #include <string.h>
+ #include <timestamp.h>
+@@ -101,6 +102,10 @@ void fsps_load(bool s3wake)
+ 	if (rdev_readat(&rdev, dest, 0, size) < 0)
+ 		die("Failed to read FSPS!\n");
+ 
++	if (IS_ENABLED(CONFIG_MEASURED_BOOT)) {
++	  tlcl_measure(1, (const void*) dest, size);
```

Comment:
Same thing here, why do you manually measure and then add the SEG_NO_MEASURE flag to prog_segment_loaded ?

Reply:
Ah - I'm wanted to measure the output before any potential changes or relocation happens (similar thing to what happens in rmodule).  I didn't want to assume much about what `fsp_component_relocate` does now or in the future.